### PR TITLE
Rdcc 2508 asb fix

### PR DESCRIPTION
**Pull request by nitishsingla1234**
### JIRA link (if applicable) ###
https://tools.hmcts.net/jira/browse/RDCC-2508

### Change description ###

Changing the implementation from Jmstemplate to Native Azure code.

**Does this PR introduce a breaking change?** (check one with "x")

```
[ ] Yes
[x] No
```

### CODE REVIEW

File: src/main/java/uk/gov/hmcts/reform/cwrdapi/CaseWorkerRefApiApplication.java [hunk 1]
```diff
@@ -6,16 +6,12 @@
...
-import org.springframework.jms.annotation.EnableJms;
-import org.springframework.retry.annotation.EnableRetry;
 import uk.gov.hmcts.reform.idam.client.IdamApi;
 
 @EnableJpaAuditing
 @EnableJpaRepositories
 @SpringBootApplication
 @EnableCaching
-@EnableJms
-@EnableRetry
```
Comment (nitishsingla1234) [inline src/main/java/uk/gov/hmcts/reform/cwrdapi/CaseWorkerRefApiApplication.java:18]: The Azure library provides a separate mechanism for the retry. I have added it in MessageConfig.java

File: src/main/java/uk/gov/hmcts/reform/cwrdapi/servicebus/TopicPublisher.java [hunk 2]
```diff
@@ -1,62 +1,115 @@
...
 
-    @Recover
-    public void recoverMessage(Exception ex) {
-        log.error("{}:: Publishing message to service bus topic failed with exception: {}:: Job Id {}",
-                loggingComponentName, ex, validationService.getAuditJobId());
-        throw new CaseworkerMessageFailedException(CaseWorkerConstants.ASB_PUBLISH_ERROR);
+    private void publishMessageToTopic(PublishCaseWorkerData caseWorkerData,
+                                      ServiceBusSenderClient serviceBusSenderClient,
+                                      ServiceBusTransactionContext transactionContext) {
+        log.info("{}:: Started publishing to topic::", loggingComponentName);
```
Comment (SujeethKumarAri) [inline src/main/java/uk/gov/hmcts/reform/cwrdapi/servicebus/TopicPublisher.java:79]: Can we add job id as well for tracking purpose ?

File: src/main/java/uk/gov/hmcts/reform/cwrdapi/servicebus/TopicPublisher.java [hunk 3]
```diff
@@ -1,62 +1,115 @@
...
+
+            // The batch is full, so we create a new batch and send the batch.
+            serviceBusSenderClient.sendMessages(messageBatch, transactionContext);
+
+            // create a new batch
+            messageBatch = serviceBusSenderClient.createMessageBatch();
+
+            // Add that message that we couldn't before.
+            if (!messageBatch.tryAddMessage(message)) {
+                log.error("Message is too large for an empty batch. Skipping. Max size: {}.",
```
Comment (SujeethKumarAri) [inline src/main/java/uk/gov/hmcts/reform/cwrdapi/servicebus/TopicPublisher.java:103]: Can we add loggingComponentName  and job id as well for tracking purpose ?

File: src/main/java/uk/gov/hmcts/reform/cwrdapi/servicebus/TopicPublisher.java [hunk 4]
```diff
@@ -1,62 +1,115 @@
...
+                        messageBatch.getMaxSizeInBytes());
+            }
+        }
+
+        if (messageBatch.getCount() > 0) {
+            serviceBusSenderClient.sendMessages(messageBatch, transactionContext);
+            log.info("Sent a batch of messages to the topic: {}", topic);
```
Comment (SujeethKumarAri) [inline src/main/java/uk/gov/hmcts/reform/cwrdapi/servicebus/TopicPublisher.java:110]: Can we add loggingComponentName  and job id as well for tracking purpose ?

File: src/main/java/uk/gov/hmcts/reform/cwrdapi/service/impl/CaseWorkerServiceImpl.java [hunk 5]
```diff
@@ -275,11 +274,8 @@ public void publishCaseWorkerDataToTopic(List<CaseWorkerProfile> caseWorkerData)
             .collect(Collectors.toUnmodifiableList());
 
         PublishCaseWorkerData publishCaseWorkerData = new PublishCaseWorkerData();
-        ListUtils.partition(caseWorkerIds, caseWorkerDataPerMessage)
-            .forEach(data -> {
-                publishCaseWorkerData.setUserIds(data);
-                topicPublisher.sendMessage(publishCaseWorkerData);
-            });
+        publishCaseWorkerData.setUserIds(caseWorkerIds);
+        topicPublisher.sendMessage(publishCaseWorkerData);
```
Comment (SujeethKumarAri) [inline src/main/java/uk/gov/hmcts/reform/cwrdapi/service/impl/CaseWorkerServiceImpl.java:278]: If you are not doing the partition here then you can just send `caseWorkerIds ` to `topicPublisher.sendMessage` like
`topicPublisher.sendMessage(caseWorkerIds);`

File: src/main/java/uk/gov/hmcts/reform/cwrdapi/config/MessagingConfig.java [hunk 6]
```diff
@@ -1,66 +1,46 @@
...
+
+    @Value("${crd.publisher.azure.service.bus.password}")
+    String sharedAccessKeyValue;
 
-    @Bean // Serialize message content to json using TextMessage
-    public MessageConverter jacksonJmsMessageConverter() {
-        MappingJackson2MessageConverter converter = new MappingJackson2MessageConverter();
-        converter.setTargetType(MessageType.TEXT);
-        return converter;
+    public ServiceBusSenderClient getServiceBusSenderClient() {
```
Comment (SujeethKumarAri) [inline src/main/java/uk/gov/hmcts/reform/cwrdapi/config/MessagingConfig.java:33]: you can create a bean for it by just appending `@Bean` on top of it

File: src/main/java/uk/gov/hmcts/reform/cwrdapi/servicebus/TopicPublisher.java [hunk 7]
```diff
@@ -1,62 +1,115 @@
...
 
-    @Retryable(maxAttempts = 3, backoff = @Backoff(delay = 2000, multiplier = 3))
-    public void sendMessage(@NotNull Object message) {
-        log.info("{}:: Publishing message to service bus topic:: Job Id is: {}", loggingComponentName,
-                validationService.getAuditJobId());
-        if (message instanceof PublishCaseWorkerData) {
+    //The Azure library has retry mechanism inbuilt, which will try to send message until timeout.
+    // There is no need for explicit retry mechanism.
+    public void sendMessage(@NotNull PublishCaseWorkerData caseWorkerIds) {
+        ServiceBusSenderClient serviceBusSenderClient = null;
```
Comment (SujeethKumarAri) [inline src/main/java/uk/gov/hmcts/reform/cwrdapi/servicebus/TopicPublisher.java:50]: you can Autowire the ServiceBusSenderClient object (once you create the  bean ServiceBusSenderClient)and use it like
`@Autowired
    ServiceBusSenderClient serviceBusSenderClient;`

File: src/main/java/uk/gov/hmcts/reform/cwrdapi/servicebus/TopicPublisher.java [hunk 8]
```diff
@@ -1,62 +1,115 @@
...
-                    validationService.getAuditJobId(),
-                    ((PublishCaseWorkerData) message).getUserIds() != null
-                            ? ((PublishCaseWorkerData) message).getUserIds().size() : null);
-        }
+                    loggingComponentName, validationService.getAuditJobId(), caseWorkerIds.getUserIds().size());
 
-        jmsTemplate.convertAndSend(destination, message);
+            serviceBusSenderClient = messagingConfig.getServiceBusSenderClient();
+            transactionContext = serviceBusSenderClient.createTransaction();
+            publishMessageToTopic(caseWorkerIds, serviceBusSenderClient, transactionContext);
```
Comment (SujeethKumarAri) [inline src/main/java/uk/gov/hmcts/reform/cwrdapi/servicebus/TopicPublisher.java:62]: no need to pass `serviceBusSenderClient ` if you Autowire it

**Review by SujeethKumarAri — CHANGES_REQUESTED** [review]

File: src/main/java/uk/gov/hmcts/reform/cwrdapi/config/MessagingConfig.java [hunk 9]
```diff
@@ -1,66 +1,46 @@
...
+
     @Value("${loggingComponentName}")
     private String loggingComponentName;
 
-    @Bean
-    public String jmsUrlString(@Value("${crd.publisher.azure.service.bus.host}") final String host) {
-        return String.format("amqps://%1s?amqp.idleTimeout=3600000", host);
-    }
+    @Value("${crd.publisher.caseWorkerDataPerMessage}")
+    private int caseWorkerDataPerMessage;
```
[the remark this answers is not in the document]
Comment (nitishsingla1234) [reply src/main/java/uk/gov/hmcts/reform/cwrdapi/config/MessagingConfig.java:19]: Agreed. It should be part of TopicPublisher class only. Removing.

File: src/main/java/uk/gov/hmcts/reform/cwrdapi/servicebus/TopicPublisher.java (hunk 2 above)
Comment (nitishsingla1234) [reply src/main/java/uk/gov/hmcts/reform/cwrdapi/servicebus/TopicPublisher.java:79]: Okay, let me add that. It's already being added in previous step also.

File: src/main/java/uk/gov/hmcts/reform/cwrdapi/servicebus/TopicPublisher.java (hunk 4 above)
Comment (nitishsingla1234) [reply src/main/java/uk/gov/hmcts/reform/cwrdapi/servicebus/TopicPublisher.java:110]: Makes sense. Adding.

File: src/main/java/uk/gov/hmcts/reform/cwrdapi/config/MessagingConfig.java (hunk 6 above)
Comment (nitishsingla1234) [reply src/main/java/uk/gov/hmcts/reform/cwrdapi/config/MessagingConfig.java:26]: Regarding the autowiring, we have deliberately avoided creating the bean.
We would like to create a fresh connection on every file upload, thus avoiding problem of long running connections.
@shashank-rastogi Can you please share your thoughts as well.

File: src/main/java/uk/gov/hmcts/reform/cwrdapi/servicebus/TopicPublisher.java (hunk 8 above)
Comment (nitishsingla1234) [reply src/main/java/uk/gov/hmcts/reform/cwrdapi/servicebus/TopicPublisher.java:62]: Kindly check the above comment.

Comment (abhijitdiwan) [reply src/main/java/uk/gov/hmcts/reform/cwrdapi/servicebus/TopicPublisher.java:60]: I think yo can Auto-wire all just keep transaction context with little life span if you want to achieve short lived transaction. I think Auto-wiring would simplifies life

Comment (nitishsingla1234) [reply src/main/java/uk/gov/hmcts/reform/cwrdapi/servicebus/TopicPublisher.java:56]: Hi @SujeethKumarAri @abhijitdiwan I have created the bean for service bus client. Can you please have a look.
Thanks

File: src/main/java/uk/gov/hmcts/reform/cwrdapi/servicebus/TopicPublisher.java (hunk 7 above)
Comment (nitishsingla1234) [reply src/main/java/uk/gov/hmcts/reform/cwrdapi/servicebus/TopicPublisher.java]: Hi @SujeethKumarAri @abhijitdiwan I have created the bean for service bus client. Can you please have a look.
Thanks

File: src/main/java/uk/gov/hmcts/reform/cwrdapi/config/MessagingConfig.java (hunk 6 above)
Comment (nitishsingla1234) [reply src/main/java/uk/gov/hmcts/reform/cwrdapi/config/MessagingConfig.java:28]: Hi @SujeethKumarAri @abhijitdiwan I have created the bean for service bus client. Can you please have a look.
Thanks

**Review by SujeethKumarAri — APPROVED** [review]

**Review by abhijitdiwan — APPROVED** [review]

**Review by SujeethKumarAri — APPROVED** [review]

**Pull request merged**